tron/commands/retry.py

```diff
@@ -0,0 +1,243 @@
+
+        self._required_action_indices = self._get_required_action_indices()
+        self._elapsed = datetime.timedelta(seconds=0)
+        self._triggers_done = False
+        self._required_actions_done = False
+        self._retry_request_result = None  # None = not issued, True = success, False = fail
+
+    @property
+    def job_run_name(self):
+        return self.full_action_name.rsplit(".", 1)[0]
+
+    @property
+    def action_name(self):
+        return self.full_action_name.rsplit(".", 1)[1]
+
+    @property
+    def status(self):
+        if not self._triggers_done:
+            return "Upstream triggers not all published"
+        elif not self._required_actions_done:
+            return "Required actions not all successfully completed"
+        elif self._retry_request_result is None:
+            return "Retry request not issued, but dependencies done"
+        elif self._retry_request_result:
+            return "Retry request issued successfully"
+        else:
+            return "Failed to issue retry request"
+
+    @property
+    def succeeded(self):
+        return bool(self._retry_request_result)
+
+    def _validate_action_name(self, full_action_name: str) -> client.TronObjectIdentifier:
+        action_run_id = client.get_object_type_from_identifier(self.tron_client.index(), full_action_name)
+        if action_run_id.type != client.TronObjectType.action_run:
+            raise ValueError(f"'{full_action_name}' is a {action_run_id.type.lower()}, not an action")
+        self.tron_client.action_runs(action_run_id.url, num_lines=0)  # verify action exists
+        return action_run_id
+
+    def _get_required_action_indices(self) -> Dict[str, bool]:
+        job_run = self.tron_client.job_runs(self.job_run_id.url)
+        required_actions = set()
+        action_indices = {}
+
+        for i, action_run in enumerate(job_run["runs"]):
+            if action_run["action_name"] == self.action_name:
+                required_actions = set(action_run["requirements"])
+            action_indices[action_run["action_name"]] = i
+
+        return dict(filter(lambda e: e[0] in required_actions, action_indices.items()))
+
+    def _log(self, s: str) -> None:
+        print(f"[{self._elapsed}] {self.full_action_name}: {s}")
+
+    async def can_retry(self) -> bool:
+        if not self._triggers_done:
+            triggers = await self.get_triggers()
+            self._triggers_done = all(triggers.values())
+            if self._triggers_done:
+                if len(triggers) > 0:
+                    self._log("All upstream triggers published")
+            else:
+                remaining_triggers = [trigger for trigger, is_done in triggers.items() if not is_done]
+                self._log(f"Upstream triggers not yet published: {remaining_triggers}")
+        if not self._required_actions_done:
+            required_actions = await self.get_required_actions()
```

Comment:
hmm, should we add a comment or rename this function to make it clear that we're not just getting the actions that need are required for a retry - we're getting the required actions that are in a "good" state for us? otherwise, someone might be tempted to cache this somewhere under the assumption that the list of required actions is static (as i was about to comment before i read more closely :p)

Reply:
Changed the name.